Add enhanced cte scheduling mode

jaystarshot opened the pull request:
## Description

Fixes #22205 

Enhance the scheduler to selectively block only the TableScan stages that depend on incomplete CTE TableWriter stages, rather than blocking all dependent sections.

Previously, all sections relying on a CTE writer were blocked until the writer completed. 
With this change, only the specific TableScans referencing the CTE are delayed, allowing other stages to proceed. This optimization can significantly improve query latency.

For example for the materialized CTE T:

```
WITH T AS (SELECT * FROM tpch.orders) 
SELECT * 
FROM T 
JOIN (SELECT * FROM customer) b 
ON t.uuid = b.uuid;

```
In this query, the right side of the join can be executed and kept in memory concurrently while the CTE write operation completes.


2 commits
1. Add the CTE used info to tablescans and table finish 
2. Use this info in scheduling by adding a central manager per scheduler and then using this to unblock/wake up tasks and to maintain completed ctes.

Can cause more resource utilization in some cases where intermediate results get blocked due to written ctes  but that may also happen without any materialization


## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
Existing UTs + prod queries

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Improve scheduling for CTE materialization: Now, only the stages containing CTE table scans that reference CTE table write stages are blocked till the write is complete, instead of the entire query being blocked as was the case previously. This is controlled by the session property ``enhanced_cte_scheduling_enabled`` (on by default) :pr:`24108`

```

jaystarshot commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSourcePartitionedScheduler.java:
```diff
@@ -179,6 +183,29 @@ public ScheduleResult schedule()
+        List<ListenableFuture<?>> blocked = new ArrayList<>();
+
+        // CTE Materialization Check
+        if (stage.requiresMaterializedCTE()) {
```
main change

jaystarshot commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -278,6 +280,17 @@ else if (state == CANCELED) {
 
         for (StageExecutionAndScheduler stageExecutionInfo : stageExecutions.values()) {
             SqlStageExecution stageExecution = stageExecutionInfo.getStageExecution();
+            // Add a listener for state changes
```
Add a listener to stages with tablefinish which updates our tracker

ZacBlanco commented on presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java:
```diff
@@ -594,6 +600,59 @@ private static Split createRemoteSplitFor(TaskId taskId, URI remoteSourceTaskLoc
+        return PlanNodeSearcher.searchFrom(source)
+                .where(planNode -> planNode instanceof TableFinishNode)
+                .findFirst()
+                .map(planNode -> ((TableFinishNode) planNode).getTemporaryTableInfo().orElseThrow(
+                        () -> new IllegalStateException("TableFinishNode has no TemporaryTableInfo")))
```
```suggestion
                .flatMap(planNode -> ((TableFinishNode) planNode).getTemporaryTableInfo())
```

jaystarshot replied:
Better to throw exp

ZacBlanco commented on presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java:
```diff
@@ -594,6 +600,59 @@ private static Split createRemoteSplitFor(TaskId taskId, URI remoteSourceTaskLoc
+    {
+        return PlanNodeSearcher.searchFrom(planFragment.getRoot())
+                .where(planNode -> (planNode instanceof TableFinishNode))
+                .findAll().stream()
```
do you think we should verify there is only one `TableFinishNode` using something like `getOnlyElement`?

rschlussel replied:
you can use findSingle() for this. Also, instead of streaming again and then anyMatch, you can move all of that into the where condition above.

ZacBlanco commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSourcePartitionedScheduler.java:
```diff
@@ -179,6 +183,29 @@ public ScheduleResult schedule()
+                    blocked.add(materializationFuture);
+                }
+            }
+            // If any CTE is not materialized, return a blocked ScheduleResult
+            if (!blocked.isEmpty()) {
+                return ScheduleResult.blocked(
+                        false,
+                        newTasks,
+                        whenAnyComplete(blocked),
```
Trying to wrap my head around the logic here:

> If any CTE is not materialized, return a blocked ScheduleResult

So the logic should be, in order to return a result which isn't blocked, then the `blocked` list should be empty. This makes sense based on the `blocked.isEmpty()` condition.

However, the `blocked` future argument is `whenAnyComplete(blocked)`. Shouldn't this be something like `whenAllComplete`? Or is there some nuance to what the `blocked` argument should be to the `ScheduleResult.blocked` function? I am not very familiar with this portion of the codebase.

jaystarshot replied:
I think any or all don't matter since blocked list just contains one future here

jaystarshot replied:
so FixedSourcePartitionedScheduler is just for the stages reading from bucketed tables. As far as I know presto doesn't allow multiple tablescans in one stage. But this code is still resilient

hantangwangd replied:
Oh yes, for hive connector, it's the bucketed tables that will expose their partitioning property into `HivePartitioningHandle` when considering colocated joins. So when left and right tables in a join have compatible bucket definition exactly on the join keys, they will be left into the same stage.

Currently, the tableScanNode for the temporary table caused by a CTE does not inherit the bucket feature of its source table, so seems that it won't stay in the same stage with other tableScanNodes. Do we have any plan to support inheriting the bucket feature of the source table in future?

rschlussel replied:
I'm still a bit confused here. A stage can have multiple CTE dependencies if they have the same bucketing, so seems like we shouldn wait for all blocked futures to complete, right? (this isn't broken now, as if we unblock when some complete, we'll just do this check again, just want to make sure i'm understanding the flow properly)

ZacBlanco commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -258,7 +259,8 @@ private SqlQueryScheduler(
 
-        this.maxConcurrentMaterializations = getMaxConcurrentMaterializations(session);
+//        this.maxConcurrentMaterializations = getMaxConcurrentMaterializations(session);
+        this.maxConcurrentMaterializations = 100;
```
remove and uncomment?

rschlussel commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/ScheduleResult.java:
```diff
@@ -57,13 +57,16 @@ public enum BlockedReason
          */
         MIXED_SPLIT_QUEUES_FULL_AND_WAITING_FOR_SOURCE,
+
+        WAITING_FOR_CTE_MATERIALIZATION,
```
add a comment explaining what this means

rschlussel commented on presto-spi/src/main/java/com/facebook/presto/spi/plan/TemporaryTableInfo.java:
```diff
@@ -0,0 +1,29 @@
+ */
+package com.facebook.presto.spi.plan;
+
+public class TemporaryTableInfo
```
add some comments explaining what this class is for

rschlussel reviewed (commented):
can you add some tests for this?

jaystarshot commented:
Thanks for the review. Actually we have a lot of good cte materialization tests which are now using this new flow [here](https://github.com/prestodb/presto/blob/e416ff11482b92f812c2c47ef5ea0f0b19d85393/presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java#L50) (over 50) so I didn’t add a new one.

hantangwangd reviewed (commented):
Thanks for the work, should we also supplement the document for configuration property `enhanced-cte-blocking`?

jaystarshot commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/CTEMaterializationTracker.java:
```diff
@@ -0,0 +1,59 @@
+        }
+        materializationFutures.put(cteName, SettableFuture.create());
+        return materializationFutures.get(cteName);
+    }
```
Yeah maybe its a good point to make it atomic, I added this in the compute() method of concurrenthashmap which is atomic

steveburnett reviewed (approved):
LGTM! (docs)

Pull branch, local doc build, looks good. Thanks!

kaikalur commented:
> In this query, the right side of the join can be executed and kept in memory concurrently while the CTE write operation completes.

Won't that increase memory for the query?

jaystarshot commented:
Yes unfortunately no way to avoid that till we read concurrently while we write maybe with a new exchange

jaystarshot commented:
@rschlussel Can you please take a look

rschlussel commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/CTEMaterializationTracker.java:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+
+    public synchronized boolean hasBeenMaterialized(String cteName)
```
why is this synchronized? don't think it should be necessary with the ConcurrentHashMap, but if it is necessary, you'd need to synchronize the writes too.

jaystarshot replied:
yes correct this is not needed

rschlussel commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/CTEMaterializationTracker.java:
```diff
@@ -0,0 +1,58 @@
+    public ListenableFuture<Void> getFutureForCTE(String cteName)
+    {
+        return materializationFutures.compute(cteName, (key, existingFuture) -> {
+            if (existingFuture == null || existingFuture.isCancelled()) {
```
why do we change isCancelled? Seems like that could be a race condition anyway (gets cancelled by another thread after we check the condition).

jaystarshot replied:
As we know, the Scheduler’s [schedule()](https://github.com/prestodb/presto/commit/baf737b12240068a54aef16a2990ee624821519e) method is single-threaded. When the CTE writer stage (or another blocked stage) becomes unblocked [here](https://github.com/prestodb/presto/blob/85cc36af2cca5be4b15d17884f3f6a01bd107781/presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java#L514), the schedule() method resumes and cancels all pending futures [here](https://github.com/prestodb/presto/blob/85cc36af2cca5be4b15d17884f3f6a01bd107781/presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java#L517). Then it will try to schedle all blocked stages again and we will call FixedSourcePartitionedScheduler’s schedule() method, at which point a cancelled future will appear in the CTEMaterializationMap.

I think you are correct when you say that a race condition can appear in some extremely unlikely case, will need to take a deeper dive on the fix here because it doesn't seem trivial. We could check and not cancel the cte futures in the schedule() method but then it would need the info passed to the future

jaystarshot replied:
I think I can add make all methods of CTEMaterializationTracker synchronized and remove all futures from the materializationFutures maps when the shedule() method resumes and cancels all pending futures [here](https://github.com/prestodb/presto/blob/85cc36af2cca5be4b15d17884f3f6a01bd107781/presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java#L517). Then it should be threadsafe.

rschlussel replied:
can we ignore the fact that it's been canceled?

jaystarshot replied:
Since our map in CTEMaterializationTracker is storing the cancelled future, the [FixedSourceCountScheduler](https://github.com/prestodb/presto/blob/85cc36af2cca5be4b15d17884f3f6a01bd107781/presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSourcePartitionedScheduler.java#L178) schedule method will return this cancelled future (which was cancelled by a previous loop of the scheduling process). I think that we need to make sure that the  FixedSourcePartitionedScheduler returns a fresh non cancelled future

jaystarshot replied:
https://github.com/prestodb/presto/pull/24108/files#diff-bbcd68b946ac8590a0a1d1371f51e32687b0061d914689eb6fb3b852b69d3ac1R534 (main change clearing all futues)

rschlussel commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/CTEMaterializationTracker.java:
```diff
@@ -0,0 +1,58 @@
+    {
+        if (materializedCtes.putIfAbsent(cteName, true) == null) {
+            SettableFuture<Void> future = materializationFutures.get(cteName);
+            if (future != null && !future.isCancelled()) {
```
same question about checking isCancelled().

rschlussel commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/CTEMaterializationTracker.java:
```diff
@@ -0,0 +1,58 @@
+/*
+ *    Tracks whether tablefinish nodes writing temporary tables for CTE Materialization are complete
+ */
+@ThreadSafe
```
I think in Presto we expect that everything is thread safe unless it's specifically noted as not thread safe.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSourcePartitionedScheduler.java:
```diff
@@ -179,6 +183,29 @@ public ScheduleResult schedule()
     {
         // schedule a task on every node in the distribution
         List<RemoteTask> newTasks = ImmutableList.of();
+        List<ListenableFuture<?>> blocked = new ArrayList<>();
```
move this inside the if(stage.requiresMaterializedCTE()) condition.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/CTEMaterializationTracker.java:
```diff
@@ -0,0 +1,68 @@
+            if (existingFuture == null) {
+                return SettableFuture.create();
+            }
+            checkArgument(!existingFuture.isCancelled(), "CTE future was found in a cancelled state");
```
this is still not thread safe, as the cancellation isn't atomic with clearing the futures, so it's possible to hit this condition (future cancelled, something tries to get the future before clearAllFutures is called).  I think it would be better to have the code consuming the futures gracefully handle receiving a cancelled future and not have to worry about synchronization.

jaystarshot replied:
Part 1 `(future canceled, something tries to get the future before clearAllFutures is called)` can be fixed by clearing all before canceling. 

For part 2 `It's also possible for the future to be canceled after it's returned, but before the consumer checks it or even before it's returned, but after this argument check` 
I think this will never happen as blocked.cancel() is invoked in SqlQueryScheduler only after all stages are scheduled. At that point, all futures have already been returned to the SqlQueryScheduler, so the consumer will have completed scheduling (and return the blocked result) before this happens.


> I think it would be better to have the code consuming the futures gracefully handle receiving a canceled future and not have to worry about synchronization.

I think the code consuming the future can be other Schedulers and making them handle canceled futures might be challenging (not 100% sure if that is even needed yet)

rschlussel replied:
that's how it works today, but it's not really thread safe, as there's no guarantee that some other thread holding the future won't cancel it or that someone won't reorder the calling code without realizing the dependency and break the assumptions that clearAllFutures is always called before cancelling any future in the map. At the very least this needs some very clear code comments about the assumptions this class is making and requirements for anything using it.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/CTEMaterializationTracker.java:
```diff
@@ -0,0 +1,70 @@
+
+    private final Map<String, Boolean> materializedCtes = new HashMap<>();
+
+    public synchronized ListenableFuture<Void> getFutureForCTE(String cteName)
```
it would be nice to avoid all this locking. it seems like right now it's only needed to make the materializationFutures and materializedCTE calls atomic (because aside from that we still make assumptions about when cancel is or is not called, so it's not providing any thread safety with regard to the cancellation calls).

  I'm thinking it would be better to use a single ConcurrentHashMap<>, and then not clear the futures before cancellation and instead check in getFutureForCTE() that it hasn't been cancelled, and if it has create a new one (like the code you had before).  something like the following.

```
private final Map<String, SettableFuture<Void>> materializationFutures = new ConcurrentHashMap<>();

public ListenableFuture<Void> getFutureForCTE(String cteName)
    {
         return materializationFutures.compute(cteName, (key, existingFuture) -> {
            if (existingFuture == null || existingFuture.iscancelled()) {
                return SettableFuture.create();
            }
            return existingFuture;
        });

   public void markCTEAsMaterialized(String cteName)
    {
         materializationFutures.compute(cteName, (key, existingFuture) -> {
            if (existingFuture == null || existingFuture.iscancelled()) {
                SettableFuture completedFuture = SettableFuture.create();
                completedFuture.set(null);
                return completedFuture;
            }
            existingFuture.set(null); // notify all listeners
            return existingFuture;
        });


```

jaystarshot replied:
Not happy with the assumptions that we are making, will brainstorm for a better way but it won't be trivial

jaystarshot replied:
```
public void markCTEAsMaterialized(String cteName)
    {
        materializationFutures.compute(cteName, (key, existingFuture) -> {
            if (existingFuture == null || existingFuture.isCancelled()) {
                SettableFuture completedFuture = SettableFuture.create();
                completedFuture.set(null);
                return completedFuture;
            }
            existingFuture.set(null); // notify all listeners
            return existingFuture;
        });
    }
```
Also here completing the future when its cancelled
```
  if (existingFuture == null || existingFuture.isCancelled()) {
                SettableFuture completedFuture = SettableFuture.create();
                completedFuture.set(null);
                return completedFuture;
            }
```
is not really useful since there will be no listeners on the new future. the entire goal of the markCTEAsMaterialized() is to unblock the stages which were waiting on this future

rschlussel replied:
the point of setting a new completed future if the original one was canceled is in order to add it to the map. that way when someone calls getFutureForCTE() they will get a future that is done.

jaystarshot replied:
Got it, first race condition is fixed.  For the second condition

> For part 2 It's also possible for the future to be canceled after it's returned, but before the consumer checks it or even before it's returned, but after this argument check

If we see the FixedSourcePartitionScheduler code 
```
 // If any CTE is not materialized, return a blocked ScheduleResult
            if (!blocked.isEmpty()) {
                return ScheduleResult.blocked(
                        false,
                        newTasks,
                        whenAnyComplete(blocked),
                        BlockedReason.WAITING_FOR_CTE_MATERIALIZATION,
                        0);
            }
```
Actually I don't think we are even cancelling the future in the CteMaterializationTracker. We are cancelling the future  which was wrapped by the `whenAnyComplete` so that problem seems non exisitent

jaystarshot replied:
Okay I think I found a good fix. If we check the [FixedSourcePatitionScheduler](https://github.com/prestodb/presto/pull/24108/files#diff-1090adf62744bbb6103f894606a0574e2d8b040779e2279e9846a820c2d6a118R200) code, we can see that it actually returns a wrapped future in whenAnyComplete. This wrapped future which can be cancelled. I also debugged and verified that if this is cancelled the inner ones are also cancelled (maybe because we cancelled with mayInterruptIfRunning?) 
However if we make sure that the ctematerialization tracker always returns a `Futures.nonCancellationPropagating` then the future in the tracker can never be cancelled so in a way we have handled cancellations. 

```
if (!blocked.isEmpty()) {
                return ScheduleResult.blocked(
                        false,
                        newTasks,
                        whenAnyComplete(blocked),
                        BlockedReason.WAITING_FOR_CTE_MATERIALIZATION,
                        0);
            }
```

jaystarshot commented:
@rschlussel can you please take a look again? Thanks

jaystarshot commented:
Rebased to resolve conflicts and added part of https://github.com/prestodb/presto/pull/21626 which was inadvertently removed. Detail [changelog](https://github.com/prestodb/presto/compare/563330c55f67791737dba82e010a2a712a05bf43..2a41b8349c4fa5792eed1a0f01ebacc38aaa3) since review
